engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -198,18 +198,18 @@ public TableLocation getTableLocationIfPresent(@NotNull final TableLocationKey t
         // See JavaDoc on tableLocations for background.
         // The intent is to create a TableLocation exactly once to replace the TableLocationKey placeholder that was
         // added in handleTableLocationKey.
-        if (current instanceof TableLocation) {
-            return (TableLocation) current;
-        }
-        synchronized (tableLocations) {
-            current = tableLocations.get(tableLocationKey);
-            if (current instanceof TableLocation) {
-                return (TableLocation) current;
+        while (!(current instanceof TableLocation)) {
+            final TableLocationKey immutableKey = (TableLocationKey) current;
+            // noinspection SynchronizationOnLocalVariableOrMethodParameter
+            synchronized (immutableKey) {
+                current = tableLocations.get(immutableKey);
+                if (immutableKey == current) {
+                    // Note, this may contend for the lock on tableLocations
+                    tableLocations.add(current = makeTableLocation(immutableKey));
```

Comment:
Should we short-circuit the instanceof / while loop check and return current here?

Reply:
We could. I thought the code was slightly more elegant this way.

Reply:
In practice, the `while` could be an `if`, I think. Unless we worry about `makeTableLocation` failing. Which could totally happen.

Reply:
Never mind, I'm wrong, `if` if better all the time, and then we don't care about short-circuiting.